### ISSUE TITLE
Sending headers on a reset HTTP/2 stream explodes.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8685]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)| @Lukasa reported|
|-|-|
|Trac ID|trac#8685|
|Type|defect|
|Created|2016-07-26 11:05:14Z|
|Branch|https://github.com/twisted/twisted/tree/8685-lukasa-sendheadersreset|

If a RST_STREAM frame is received before headers are written, we can hit a crazy traceback:

```
Traceback (most recent call last):
  File "/Users/cory/tmp/twisted/twisted/web/test/test_http2.py", line 2082, in test_dataOnClose
    a.dataReceived(requestBytes)
  File "/Users/cory/tmp/twisted/twisted/web/_http2.py", line 170, in dataReceived
    self._requestEnded(event)
  File "/Users/cory/tmp/twisted/twisted/web/_http2.py", line 443, in _requestEnded
    stream.requestComplete()
  File "/Users/cory/tmp/twisted/twisted/web/_http2.py", line 917, in requestComplete
    self._request.requestReceived(self.command, self.path, b'HTTP/2')
  File "/Users/cory/tmp/twisted/twisted/web/http.py", line 766, in requestReceived
    self.process()
  File "/Users/cory/tmp/twisted/twisted/web/test/test_http.py", line 62, in process
    self.write(request)
  File "/Users/cory/tmp/twisted/twisted/web/http.py", line 940, in write
    self.channel.writeHeaders(version, code, reason, headers)
  File "/Users/cory/tmp/twisted/twisted/web/_http2.py", line 990, in writeHeaders
    self._conn.writeHeaders(version, code, reason, headers, self.streamID)
  File "/Users/cory/tmp/twisted/twisted/web/_http2.py", line 508, in writeHeaders
    self.conn.send_headers(streamID, headers)
  File "/Users/cory/tmp/twisted/build/lib/python2.7/site-packages/h2/connection.py", line 668, in send_headers
    headers, self.encoder, end_stream
  File "/Users/cory/tmp/twisted/build/lib/python2.7/site-packages/h2/stream.py", line 756, in send_headers
    events = self.state_machine.process_input(input_)
  File "/Users/cory/tmp/twisted/build/lib/python2.7/site-packages/h2/stream.py", line 118, in process_input
    return func(self, previous_state)
  File "/Users/cory/tmp/twisted/build/lib/python2.7/site-packages/h2/stream.py", line 298, in send_on_closed_stream
    raise StreamClosedError(self.stream_id)
h2.exceptions.StreamClosedError: 
```

In this instance, we should probably handle this better. Or, you know, at all.

<details><summary>Searchable metadata</summary>

```
trac-id__8685 8685
type__defect defect
reporter__Lukasa Lukasa
priority__normal normal
milestone__None None
branch__8685_lukasa_sendheadersreset 8685-lukasa-sendheadersreset
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1469531114104602 1469531114104602
changetime__1469804469198641 1469804469198641
version__None None
owner__Lukasa Lukasa

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_1"></a>|@Lukasa commented|
|-|-|

In changeset f3b34d96bed245ad7ef96e8eadf50d6f2dc788d9
```
#!CommitTicketReference repository="" revision="f3b34d96bed245ad7ef96e8eadf50d6f2dc788d9"
Failing test for #8685.
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_2"></a>|@Lukasa commented|
|-|-|

In changeset c19710d78f962912ac1c9ebc8c9081ee569e000b
```
#!CommitTicketReference repository="" revision="c19710d78f962912ac1c9ebc8c9081ee569e000b"
Topfile for #8685.
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_4"></a>|@Lukasa commented|
|-|-|

Patch available in https://github.com/twisted/twisted/pull/418

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_5"></a>|@Lukasa commented|
|-|-|

In changeset b647e6229e7b9fc5269601d0fdf30450b8df8b84
```
#!CommitTicketReference repository="" revision="b647e6229e7b9fc5269601d0fdf30450b8df8b84"
Failing test for #8685.
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_6"></a>|@Lukasa commented|
|-|-|

In changeset b79f469dcd152336bdfa9df8fc21d61e7af73e3f
```
#!CommitTicketReference repository="" revision="b79f469dcd152336bdfa9df8fc21d61e7af73e3f"
Topfile for #8685.
```

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set owner to @Lukasa|
|-|-|

Changes looks good.

I left some minor inline comments in the review.

I feel that the current `buildRequestBytes` name is not a good DSL :) as it serializes 2 frames ... last frame with a end of stream marker 

Please check my comments and merge at will... as my comments are based on feeling and not facts :)

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|GitHub set status to `closed`|
|-|-|

In changeset bb30eb04c71e1158d0cb381aa347657fbce4d113
```
#!CommitTicketReference repository="" revision="bb30eb04c71e1158d0cb381aa347657fbce4d113"
Merge 8685-lukasa-sendheadersreset: Don't explode by sending headers on a reset stream.

Author: Lukasa
Reviewer: adiroiban
Fixes: #8685
```
